packages/cogify/src/tile.cover.ts

```diff
@@ -0,0 +1,201 @@
+import { ConfigImageryTiff } from '@basemaps/config/build/json/tiff.config.js';
+import { BoundingBox, Bounds, EpsgCode, Projection, QuadKey, Simplify, TileId, TileMatrixSet } from '@basemaps/geo';
+import { LogType } from '@basemaps/shared';
+import { CliInfo } from '@basemaps/shared/build/cli/info.js';
+import { intersection, MultiPolygon, toFeatureCollection, union } from '@linzjs/geojson';
+import { Metrics } from '@linzjs/metrics';
+import { createCovering } from './cogify/covering.js';
+import { CogifyDefaults } from './cogify/gdal.js';
+import { CogifyLinkCutline, CogifyLinkSource, CogifyStacCollection, CogifyStacItem } from './cogify/stac.js';
+import { CutlineOptimizer } from './cutline.js';
+
+export interface TileCoverContext {
+  /** Unique id for the covering */
+  id: string;
+  /** List of imagery to cover */
+  imagery: ConfigImageryTiff;
+  /** Cutline to apply */
+  cutline: CutlineOptimizer;
+  /** Output tile matrix */
+  tileMatrix: TileMatrixSet;
+  /** Optional metrics provider to track how long actions take */
+  metrics?: Metrics;
+  /** Optional logger to trace covering creation */
+  logger?: LogType;
+}
+export interface TileCoverResult {
+  /** Stac collection for the imagery */
+  collection: CogifyStacCollection;
+  /** tiles to create */
+  items: CogifyStacItem[];
+  /** GeoJSON features of all the source imagery */
+  source: GeoJSON.FeatureCollection;
+}
+
+export async function createTileCover(ctx: TileCoverContext): Promise<TileCoverResult> {
+  // Find the zoom level that is at least as good as the source imagery
+  const targetBaseZoom = Projection.getTiffResZoom(ctx.tileMatrix, ctx.imagery.gsd);
+  const targetPixelScale = ctx.tileMatrix.pixelScale(targetBaseZoom);
+
+  // The base zoom is 256x256 pixels at its resolution, we are trying to find a image that is <32k pixels wide/high
+  // zooming out 7 levels converts a 256x256 image into 32k x 32k image
+  // 256 * 2 ** 7 = 32,768 - 256x256 tile
+  // 512 * 2 ** 6 = 32,768 - 512x512 tile
+  const optimalCoveringZoom = targetBaseZoom - 7; // z12 from z19
```

Comment:
Is this possible become less then 0? Like satellite imagery with a small `targetBaseZoom`. Do we want to constraint this number to be larger than 10 or some other number that not too large?

Reply:
Yes we do, should limit to `Math.min(1, ...)` good :eyes: